File: test_runner.c
#include <stdio.h>
#include <time.h>

int recursive_fibonacci(int n) {
    if (n == 0) return 0;
    if (n == 1) return 1;
    return recursive_fibonacci(n-1) + recursive_fibonacci(n-2);
}

int iterative_fibonacci(int n) {
    int fib[n+1], i;
    fib[0] = 0;
    fib[1] = 1;
    for (i = 2; i <= n; i++)
        fib[i] = fib[i-1] + fib[i-2];
    return fib[n];
}

int dynamic_fibonacci(int n) {
    int F[2][2] = {{1,1},{1,0}};
    if (n == 0)
        return 0;
    power(F, n-1);
    return F[0][0];
}

void multiply(int F[2][2], int M[2][2]) {
    int x =  F[0][0]*M[0][0] + F[0][1]*M[1][0];
    int y =  F[0][0]*M[0][1] + F[0][1]*M[1][1];
    int z =  F[1][0]*M[0][0] + F[1][1]*M[1][0];
    int w =  F[1][0]*M[0][1] + F[1][1]*M[1][1];

    F[0][0] = x;
    F[0][1] = y;
    F[1][0] = z;
    F[1][1] = w;
}

void power(int F[2][2], int n) {
    int i;
    int M[2][2] = {{1,1},{1,0}};

    for (i = 2; i <= n; i++)
        multiply(F, M);
}

double get_execution_time(int n, int algorithm) {
    clock_t start, end;
    double time_spent;
    start = clock();

    switch(algorithm) {
        case 0:
            recursive_fibonacci(n);
            break;
        case 1:
            iterative_fibonacci(n);
            break;
        case 2:
            dynamic_fibonacci(n);
            break;
    }

    end = clock();
    time_spent = (double)(end - start) / CLOCKS_PER_SEC;
    return time_spent;
}

void print_table(int n) {
    int i;
    printf("| n | Recursive C | Iterative C | Dynamic Programming C |\n");
    printf("|---|-------------|-------------|------------------------|\n");
    for (i = 1; i <= n; i++) {
        printf("| %d | %11.5f | %11.5f | %11.5f |\n", i, get_execution_time(i, 0), get_execution_time(i, 1), get_execution_time(i, 2));
    }
}

int main(int argc, char* argv[]) {
    int n, i;
    if (argc > 1) {
        n = atoi(argv[1]);
    } else {
        n = 30;
    }
    print_table(n);
    return 0;
}

File: test_runner.py
import subprocess
import time
import sys
import math
from typing import Union

COMMON_ARG_FORMAT = "python3 Fibonacci.py {n} {algo}"
FORMAT = "markdown"
TIMEOUT = 60

LAST_RUN_TRACKER = {"0": 0.0, "1": 0.0, "2": 0.0}


def run_single(n: int, typ: Union[int, str], command=COMMON_ARG_FORMAT) -> float:
    """Run a single instance collecting the total execution time

    Args:
        command:
        n (int): the Fibonacci number to compute
        typ (int): the type of algorithm to use

    Returns:
        float: the time it took, or nan if TIMEOUT is reached first
    """
    command = command.format(n=n, algo=typ)
    if math.isnan(LAST_RUN_TRACKER[str(typ)]):
        return math.nan  # skip running if we are already timing out
    try:
        start = time.time()
        subprocess.run(command.split(), timeout=TIMEOUT)
        end = time.time()
        result = end - start

    except subprocess.TimeoutExpired:
        result = math.nan
    LAST_RUN_TRACKER[str(typ)] = result
    return result


def build_row(n: int) -> str:
    """Builds a row to print to the screen either in csv format or markdown

    Args:
        n (int): the Fibonacci number to compute

    Returns:
        str: a markdown formatted or csv string of the result
    """
    results_lst = []
    for t in range(0, 3):
        result = run_single(n, t)
        results_lst.append("-" if math.isnan(result) else f"{result:.5f}")
    recursive, iterative, dynamic_Programming = results_lst

    if FORMAT == "markdown":
        return f"| {n:<4} | {recursive.center(8, ' ')} | {iterative.center(8, ' ')} | {dynamic_Programming.center(8, ' ')} |"
    return f"{n},{recursive},{iterative},{dynamic_Programming}"


def table_header() -> str:
    """Returns a markdown table header for this data set"""
    if FORMAT == "markdown":
        return "| n | Recursive P | Iterative P | Dynamic Programming P |\n" + \
            "|--|:--:|:--:|:--:|"
    return "n,Recursive P,Iterative P,Dynamic Programming P"


def main(n):
    print(table_header())
    for i in range(n + 1):
        print(build_row(i))


# note while using argv directly, there are better tools for this like pip click
if __name__ == "__main__":
    _n = 30 if len(sys.argv) < 2 else int(sys.argv[1])
    if len(sys.argv) == 3:
        FORMAT = "csv"
    main(_n)
